swap/simulations_test.go

```diff
@@ -0,0 +1,695 @@
+// To do this, we pass a `swapSimulationParams` object to this function,
+// which contains the shared objects needed to initialize the `SimulatedBackend`
+func newSimServiceMap(params *swapSimulationParams) map[string]simulation.ServiceFunc {
+	simServiceMap := map[string]simulation.ServiceFunc{
+		// we use a swap service
+		"swap": func(ctx *adapters.ServiceContext, bucket *sync.Map) (s node.Service, cleanup func(), err error) {
+			// every simulation node has an instance of a `testService`
+			ts := newTestService()
+			// balance is the interface for `NewAccounting`; it is a Swap
+			balance := params.swaps[params.count]
+			dir := params.dirs[params.count]
+			// every node is a different instance of a Swap and gets a different entry in the map
+			params.count++
+			ts.spec = newTestSpec()
+			// create the accounting instance and assign to the spec
+			ts.spec.Hook = protocols.NewAccounting(balance)
+			ts.swap = balance
+			// deploy the accounting to the `SimulatedBackend`
+			err = testDeploy(context.Background(), balance)
+			if err != nil {
+				return nil, nil, err
+			}
+			// store the testService into the bucket
+			bucket.Store(bucketKeySwap, ts)
+
+			cleanup = func() {
+				ts.swap.store.Close()
+				os.RemoveAll(dir)
+			}
+
+			return ts, cleanup, nil
+		},
+	}
+	return simServiceMap
+}
+
+// newSharedBackendSwaps pre-loads each simulated node account with "funds"
+// so that later in the simulation all operations have sufficient gas
+func newSharedBackendSwaps(nodeCount int) (*swapSimulationParams, error) {
+	params := &swapSimulationParams{
+		swaps:       make(map[int]*Swap),
+		dirs:        make(map[int]string),
+		maxMsgPrice: 10000,
+		minMsgPrice: 100,
+		nodeCount:   nodeCount,
+	}
+	keys := make(map[int]*ecdsa.PrivateKey)
+	addrs := make(map[int]common.Address)
+	alloc := core.GenesisAlloc{}
+	stores := make(map[int]*state.DBStore)
+
+	// for each node, generate keys, a GenesisAccount and a state store
+	for i := 0; i < nodeCount; i++ {
+		key, err := crypto.GenerateKey()
+		if err != nil {
+			return nil, err
+		}
+		keys[i] = key
+		addrs[i] = crypto.PubkeyToAddress(key.PublicKey)
+		alloc[addrs[i]] = core.GenesisAccount{Balance: big.NewInt(10000000000)}
+		dir, err := ioutil.TempDir("", fmt.Sprintf("swap_test_store_%x", addrs[i].Hex()))
+		if err != nil {
+			return nil, err
+		}
+		stateStore, err := state.NewDBStore(dir)
+		if err != nil {
+			return nil, err
+		}
+		params.dirs[i] = dir
+		stores[i] = stateStore
+	}
+	// then create the single SimulatedBackend
+	gasLimit := uint64(8000000000)
+	defaultBackend := backends.NewSimulatedBackend(alloc, gasLimit)
+	testBackend := &swapTestBackend{SimulatedBackend: defaultBackend}
+	// finally, create all Swap instances for each node, which share the same backend
+	var owner *Owner
+	defParams := newDefaultParams()
+	for i := 0; i < nodeCount; i++ {
+		owner = createOwner(keys[i])
+		params.swaps[i] = new(stores[i], owner, testBackend, defParams)
```

Comment:
In fact I realized this while working on this PR, there was a `new` call for a pointer creation (which is now removed), and it was resolving to the package `new` call instead of the builtin...so I noted this and we will change this in a future PR